File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        print("iteration")
        if root is None:
            print("None")
            return 0
        else:
            print(root.val)
            l = self.maxDepth(root.left)
            r = self.maxDepth(root.right)
            print(l, " ", r)
        return max(l, r)+1

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
int help_function(TreeNode* root, std::set<int, std::greater<int>>& s1, int biggest_so_far)
    {
        if(root == nullptr)return biggest_so_far;
        int max_l = help_function(root->left, s1, biggest_so_far);
        int max_r = help_function(root->right, s1, biggest_so_far);
        if(root->val + max_l < root->val) max_l =0;
        if(root->val + max_r < root->val) max_r =0;

        s1.insert(max_l + max_r + root->val);

        return std::max(max_l, max_r)+ root->val;
    }
  
    
    int maxPathSum(TreeNode* root) {
        if(root == nullptr)return 0;
        std::set<int, std::greater<int>> s1{};
        int val = help_function(root, s1, -1001);
        return max(*s1.begin(), val); 
    }
};
